clang/lib/Sema/HLSLExternalSemaSource.cpp

```diff
@@ -323,30 +324,99 @@ struct TemplateParameterListBuilder {
         S.Context, Builder.Record->getDeclContext(), SourceLocation(),
         SourceLocation(), /* TemplateDepth */ 0, Position,
         &S.Context.Idents.get(Name, tok::TokenKind::identifier),
-        /* Typename */ false,
-        /* ParameterPack */ false);
+        /* Typename */ true,
+        /* ParameterPack */ false,
+        /* HasTypeConstraint*/ false);
     if (!DefaultValue.isNull())
       Decl->setDefaultArgument(
           S.Context, S.getTrivialTemplateArgumentLoc(DefaultValue, QualType(),
                                                      SourceLocation()));
-
+    Decl->setReferenced();
     Params.emplace_back(Decl);
     return *this;
   }
 
-  BuiltinTypeDeclBuilder &finalizeTemplateArgs() {
+  ConceptSpecializationExpr *getConceptSpecializationExpr(Sema &S,
+                                                          ConceptDecl *CD) {
+    ASTContext &context = S.getASTContext();
+    SourceLocation loc = Builder.Record->getBeginLoc();
+    DeclarationNameInfo DNI(CD->getDeclName(), loc);
+    NestedNameSpecifierLoc NNSLoc;
+    DeclContext *DC = Builder.Record->getDeclContext();
+    TemplateArgumentListInfo TALI(loc, loc);
+
+    // assume that the concept decl has just one template parameter
+    TemplateTypeParmDecl *ConceptTTPD = dyn_cast<TemplateTypeParmDecl>(
+        CD->getTemplateParameters()->getParam(0));
+
+    clang::TemplateTypeParmDecl *T = clang::TemplateTypeParmDecl::Create(
```

Comment:
I wonder if some psuedo-c++ code in a comment showing what this these AST nodes represent would help someone reading this?

Reply:
I've added some helpful comments.

Reply:
Would it not be meaningful to show the C++ code that generated this AST?  I think that would help me understand it all better, but maybe if I saw it I'd realize I was wrong! :)

Reply:
Added! I've left out the expected AST structure since it probably won't ever conform to clang format.